### PR TITLE
feat(formats): output references handles interpoloated references

### PR DESCRIPTION
**Pull request by dbanksdesign**
*Issue #, if available:* #589

*Description of changes:*

* Organized formatHelpers into their own files
* Renamed `getReference` to `getReferences` and changed it to return an array of references for cases when there are multiple references in a single value.
* Renamed `sortAllProperties` format helper to `sortByReference` to be more clear and made changes to support interpolated values and to not mutate the `allProperties` array. This caused some of the sorting changes in some of the snapshots. 
* Added `createPropertyFormatter` format helper which will return a formatting function based on some configuration to be used in formats when iterating over `allProperties`. The `formattedVariables` format helper uses this function as well as several formats to remove duplicate code.
* Updated some formats to use the new format helpers
* Updated the `flutter/dart.class` to use the sorting when using `outputReferences`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/formats.md [hunk 1]
```diff
@@ -240,7 +240,7 @@ The formatter function that is called when Style Dictionary builds files.
     <td>args.dictionary.usesReference</td><td><code>function</code></td><td><p>Use this function to see if a token&#39;s value uses a reference. This is the same function style dictionary uses internally to detect a reference.</p>
 </td>
     </tr><tr>
-    <td>args.dictionary.getReference</td><td><code>function</code></td><td><p>Use this function to get the token/property that it references. You can use this to output a reference in your custom format. For example: <code>dictionary.getReference(token.original.value) // returns the referenced token object</code></p>
+    <td>args.dictionary.getReferences</td><td><code>function</code></td><td><p>Use this function to get the token/property that it references. You can use this to output a reference in your custom format. For example: <code>dictionary.getReferences(token.original.value) // returns an array of the referenced token objects</code></p>
```
Comment (dbanksdesign) [inline docs/formats.md:243]: typo: tokens/properties

File: lib/common/formatHelpers/formattedVariables.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  if (outputReferences) {
+    // note: using the spread operator here so we get a new array rather than
+    // mutating the original
+    allProperties = [...allProperties].sort(sortByReference(dictionary));
+  }
+
+  return allProperties
+    .map(createPropertyFormatter({ outputReferences, dictionary, format }))
+    .filter(function(strVal) { return !!strVal })
+    .join('\n');
```
Comment (chazzmoney) [inline lib/common/formatHelpers/formattedVariables.js:53]: might want to make the line ending character a parameter with a default of \n

File: lib/register/format.js [hunk 3]
```diff
@@ -25,7 +25,7 @@
  * @param {Object} args.dictionary.properties - Object structure of the tokens/properties that has been transformed and references resolved.
  * @param {Array} args.dictionary.allProperties - Flattened array of all the tokens/properties. This makes it easy to output a list, like a list of SCSS variables.
  * @param {function(value): Boolean} args.dictionary.usesReference - Use this function to see if a token's value uses a reference. This is the same function style dictionary uses internally to detect a reference.
- * @param {function(value): Value} args.dictionary.getReference - Use this function to get the token/property that it references. You can use this to output a reference in your custom format. For example: `dictionary.getReference(token.original.value) // returns the referenced token object`
+ * @param {function(value): Value} args.dictionary.getReferences - Use this function to get the token/property that it references. You can use this to output a reference in your custom format. For example: `dictionary.getReferences(token.original.value) // returns an array of the referenced token objects`
```
Comment (chazzmoney) [inline lib/register/format.js:28]: change documentation to match above as per our conversation

File: test.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      files: [{
+        destination: `variables.css`,
+        format: `css/variables`,
+        options: {
+          outputReferences: true
+        }
+      }]
+    }
+  }
+}).buildAllPlatforms();
```
Comment (chazzmoney) [inline test.js:24]: oopsie

File: variables.css [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+ * Do not edit directly
+ * Generated on Tue, 30 Mar 2021 19:49:14 GMT
+ */
+
+:root {
+  --color-blue: #0000ff;
+  --color-red: #ff0000;
+  --color-test: var(--color-red);
+  --color-danger: calc(var(--color-red), var(--color-blue));
+}
```
Comment (chazzmoney) [inline variables.css:11]: oopsie

File: scripts/handlebars/templates/formats.hbs [hunk 6]
```diff
@@ -252,14 +252,14 @@ StyleDictionary.registerFormat({
     return dictionary.allProperties.map(token => {
       let value = JSON.stringify(token.value);
       // the `dictionary` object now has `usesReference()` and
-      // `getReference()` methods. `usesReference()` will return true if
-      // the value has a reference in it. `getReference()` will return
+      // `getReferences()` methods. `usesReference()` will return true if
+      // the value has a reference in it. `getReferences()` will return
       // the reference to the whole token so that you can access its
```
Comment (chazzmoney) [inline scripts/handlebars/templates/formats.hbs:257]: > reference to the whole token

File: lib/utils/references/getReferences.js [hunk 7]
```diff
@@ -30,30 +30,34 @@ const GroupMessages = require('../groupMessages');
...
       ref = resolveReference(pathName, self._properties);
       // and warn the user about this
       GroupMessages.add(GroupMessages.GROUP.FilteredOutputReferences, variable);
     }
-
+    references.push(ref);
   });
 
-  return ref;
+  return references;
```
Comment (chazzmoney) [inline lib/utils/references/getReferences.js:60]: 🧠

File: lib/common/formatHelpers/fileHeader.js [hunk 8]
```diff
@@ -55,26 +64,27 @@ function fileHeader(file, commentStyle) {
...
+    header = `/**${lineSeparator}`;
+    footer = `${lineSeparator} */${lineSeparator}${lineSeparator}`;
   }
 
-  return `${commentHeader}${fn(defaultHeader)
-    .map(line => `${commentPrefix}${line}`)
-    .join('\n')}${commentFooter}`;
+  return `${header}${fn(defaultHeader)
+    .map(line => `${prefix}${line}`)
+    .join(lineSeparator)}${footer}`;
```
Comment (chazzmoney) [inline lib/common/formatHelpers/fileHeader.js]: ❤️

**Pull request merged**